File: Arrays/IntervalsBetweenIdenticalElements.go
// LeetCode 2121 - Intervals Between Identical Elements
// Time Complexity: O(n) | Space Complexity: O(n)
package main

func getDistances(arr []int) []int64 {
	n := len(arr)
	ans := make([]int64, n)
	groups := make(map[int][]int)

	for i, value := range arr {
		groups[value] = append(groups[value], i)
	}

	for _, indices := range groups {
		m := len(indices)
		left := int64(0)
		right := int64(-m * indices[0])

		for _, index := range indices {
			right += int64(index)
		}

		for i, index := range indices {
			ans[index] = left + right

			if i+1 < m {
				gap := int64(indices[i+1] - index)
				left += gap * int64(i+1)
				right -= gap * int64(m-i-1)
			}
		}
	}

	return ans
}

File: Arrays/IntervalsBetweenIdenticalElements.java
// LeetCode 2121 - Intervals Between Identical Elements
// Time Complexity: O(n) | Space Complexity: O(n)
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class IntervalsBetweenIdenticalElements {
    public long[] getDistances(int[] arr) {
        int n = arr.length;
        long[] ans = new long[n];
        Map<Integer, List<Integer>> groups = new HashMap<>();

        for (int i = 0; i < n; i++) {
            groups.computeIfAbsent(arr[i], key -> new ArrayList<>()).add(i);
        }

        for (List<Integer> indices : groups.values()) {
            int m = indices.size();
            long left = 0;
            long right = -1L * m * indices.get(0);

            for (int index : indices) {
                right += index;
            }

            for (int i = 0; i < m; i++) {
                ans[indices.get(i)] = left + right;

                if (i + 1 < m) {
                    long gap = indices.get(i + 1) - indices.get(i);
                    left += gap * (i + 1L);
                    right -= gap * (m - i - 1L);
                }
            }
        }

        return ans;
    }
}
